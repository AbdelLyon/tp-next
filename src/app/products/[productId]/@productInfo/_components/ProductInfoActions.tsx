"use client";
import { Button } from "@/components/ui/button";
import { ShoppingCart, ArrowRight } from "lucide-react";

export const ProductInfoActions = () => {
  return (
    <div className="grid grid-cols-2 gap-4 mt-8" data-testid="product-actions">
      <Button
        size="lg"
        className="font-medium bg-primary hover:opacity-80 shadow-sm hover:shadow transition-all duration-300 rounded-lg"
        data-testid="add-to-cart-button"
        onClick={() => console.log("Add to cart button clicked")}
      >
        <ShoppingCart className="mr-2 h-4 w-4" />
        Add to Cart
      </Button>
      <Button
        size="lg"
        variant="outline"
        className="font-medium border-primary text-primary hover:opacity-80 transition-colors duration-300 rounded-lg"
        data-testid="buy-now-button"
        onClick={() => console.log("Buy now button clicked")}
      >
        Buy Now <ArrowRight className="ml-2 h-4 w-4" />
      </Button>
    </div>
  );
};
