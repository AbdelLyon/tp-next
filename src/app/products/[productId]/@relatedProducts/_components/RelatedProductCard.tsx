import Image from "next/image";
import Link from "next/link";
import { Card } from "@/components/ui/card";
import { Badge } from "@/components/ui/badge";
import { Skeleton } from "@/components/ui/skeleton";
import { ProductModel } from "@/types/product";
import { Star } from "lucide-react";
import { mergeTailwindClasses } from "x-react/utils";

interface RelatedProductCardProps {
  product: ProductModel;
}

export const RelatedProductCard = ({ product }: RelatedProductCardProps) => {
  const originalPrice = product.discountPercentage
    ? product.price * (1 + product.discountPercentage / 100)
    : null;

  return (
    <Link
      href={`/products/${product.id}`}
      className="block"
      data-testid={`related-product-${product.id}`}
    >
      <Card className="overflow-hidden hover:shadow-lg transition-all duration-300 group border border-border bg-content1 hover:border-primary/20">
        <div className="flex items-center gap-4 p-3">
          <ProductThumbnail
            thumbnail={product.thumbnail}
            title={product.title}
            discountPercentage={product.discountPercentage}
          />

          <div className="flex-1 min-w-0">
            <h3
              className="font-medium text-sm line-clamp-1"
              data-testid="product-title"
            >
              {product.title}
            </h3>

            {product.brand && (
              <p
                className="text-xs text-muted-foreground line-clamp-1 mt-0.5"
                data-testid="product-brand"
              >
                {product.brand}
              </p>
            )}

            <ProductRating rating={product.rating} />

            <div className="flex items-center justify-between mt-2">
              <ProductPrice
                price={product.price}
                originalPrice={originalPrice}
              />

              <Badge
                variant="outline"
                className="text-xs bg-primary/5 text-primary border-border group-hover:bg-primary/10 transition-colors duration-300"
                data-testid="view-badge"
              >
                View
              </Badge>
            </div>
          </div>
        </div>
      </Card>
    </Link>
  );
};

const ProductThumbnail = ({
  thumbnail,
  title,
  discountPercentage,
}: {
  thumbnail?: string;
  title: string;
  discountPercentage?: number;
}) => (
  <div
    className="relative h-20 w-20 flex-shrink-0 rounded-md overflow-hidden bg-content-2"
    data-testid="product-thumbnail"
  >
    {thumbnail ? (
      <Image
        src={thumbnail}
        alt={title}
        fill
        className="object-cover transition-transform duration-500 group-hover:scale-110"
        sizes="80px"
        unoptimized={true}
      />
    ) : (
      <Skeleton className="h-full w-full" />
    )}
    {discountPercentage && (
      <div
        className="absolute top-0 right-0 bg-red-500 text-xs font-bold px-1.5 py-0.5 rounded-bl-md"
        data-testid="discount-badge"
      >
        -{discountPercentage.toFixed(0)}%
      </div>
    )}
  </div>
);

const ProductRating = ({ rating }: { rating?: number }) => (
  <>
    {rating && (
      <div
        className="flex items-center gap-0.5 mt-1.5"
        data-testid="product-rating"
      >
        {[...Array(5)].map((_, i) => (
          <Star
            key={i}
            className={mergeTailwindClasses(
              "h-3 w-3",
              i < Math.floor(rating ?? 0)
                ? "text-amber-400 fill-amber-400"
                : i < rating
                ? "text-amber-400 fill-amber-400 opacity-50"
                : "",
            )}
            data-testid={`star-${i}`}
          />
        ))}
        <span className="text-xs ml-1">{rating.toFixed(1)}</span>
      </div>
    )}
  </>
);

const ProductPrice = ({
  price,
  originalPrice,
}: {
  price: number;
  originalPrice: number | null;
}) => (
  <div className="flex items-baseline gap-1.5" data-testid="product-price">
    <span className="font-medium text-sm" data-testid="current-price">
      ${price.toFixed(2)}
    </span>
    {originalPrice && (
      <span className="text-xs line-through" data-testid="original-price">
        ${originalPrice.toFixed(2)}
      </span>
    )}
  </div>
);
