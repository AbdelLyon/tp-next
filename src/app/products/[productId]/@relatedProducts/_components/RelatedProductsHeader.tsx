import Link from "next/link";
import { ReactNode } from "react";

interface RelatedProductsHeaderProps {
  title: string;
  actionLink?: string;
  actionText?: string;
  icon?: ReactNode;
}

export const RelatedProductsHeader = ({
  title,
  actionLink,
  actionText,
  icon,
}: RelatedProductsHeaderProps) => {
  return (
    <div
      className="flex items-center justify-between mb-5"
      data-testid="section-header"
    >
      <h2
        className="text-xl font-semibold bg-gradient-to-r from-primary to-primary/80 bg-clip-text text-transparent"
        data-testid="section-title"
      >
        {title}
      </h2>

      {actionLink && actionText && (
        <Link
          href={actionLink}
          className="text-sm text-primary/80 hover:text-primary flex items-center gap-1 transition-colors"
          data-testid="section-action-link"
        >
          {actionText} {icon}
        </Link>
      )}
    </div>
  );
};
