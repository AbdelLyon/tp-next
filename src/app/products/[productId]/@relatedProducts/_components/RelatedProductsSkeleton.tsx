import { Card } from "@/components/ui/card";
import { Skeleton } from "@/components/ui/skeleton";
import { Separator } from "@/components/ui/separator";

export const RelatedProductsSkeleton = () => {
  return (
    <div className="bg-white rounded-lg border shadow-sm p-6">
      <Skeleton className="h-7 w-48 mb-4" />

      <div className="space-y-4">
        {Array(4)
          .fill(0)
          .map((_, index) => (
            <Card key={index} className="overflow-hidden">
              <div className="flex items-center gap-4 p-3">
                <div className="relative h-16 w-16 flex-shrink-0 rounded-md overflow-hidden bg-gray-100">
                  <Skeleton className="h-full w-full" />
                </div>

                <div className="flex-1 min-w-0">
                  <Skeleton className="h-4 w-3/4 mb-2" />
                  <Skeleton className="h-3 w-1/2 mb-2" />
                  <div className="flex items-center justify-between mt-1">
                    <Skeleton className="h-4 w-16" />
                    <Skeleton className="h-5 w-12 rounded-full" />
                  </div>
                </div>
              </div>
            </Card>
          ))}
      </div>

      <Separator className="my-6" />

      <Skeleton className="h-5 w-36 mb-3" />
      <div className="flex flex-wrap gap-2">
        {Array(4)
          .fill(0)
          .map((_, index) => (
            <Skeleton key={index} className="h-6 w-20 rounded-full" />
          ))}
      </div>
    </div>
  );
};
