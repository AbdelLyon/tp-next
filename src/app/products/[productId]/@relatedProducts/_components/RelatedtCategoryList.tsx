// CategoryList.tsx
import { Badge } from "@/components/ui/badge";

interface RelatedtCategoryListProps {
  title: string;
  categories: string[];
}

export const RelatedtCategoryList = ({
  title,
  categories,
}: RelatedtCategoryListProps) => {
  return (
    <div className="flex flex-col space-y-3" data-testid="category-list">
      <h3 className="text-sm font-medium" data-testid="category-list-title">
        {title}
      </h3>
      <div className="flex flex-wrap gap-2">
        {categories.map((category) => (
          <Badge
            key={category}
            variant="secondary"
            className="cursor-pointer hover:bg-primary/10 hover:text-primary transition-all duration-300 py-1.5"
            data-testid={`category-${category
              .toLowerCase()
              .replace(/\s+/g, "-")}`}
          >
            {category}
          </Badge>
        ))}
      </div>
    </div>
  );
};
