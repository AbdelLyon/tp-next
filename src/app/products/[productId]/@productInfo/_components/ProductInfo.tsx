import { productService } from "@/services/ProductService";
import { ProductInfoGallery } from "./ProductInfoGallery";
import { ProductInfoHeader } from "./ProductInfoHeader";
import { ProductInfoPricing } from "./ProductInfoPricing";
import { ProductInfoDetails } from "./ProductInfoDetails";
import { ProductInfoActions } from "./ProductInfoActions";
import { ProductInfoShipping } from "./ProductInfoShipping";
import { ProductInfoTabs } from "./ProductInfoTabs";

const ProductInfo = async ({ productId }: { productId: string }) => {
  const product = await productService.getProductById(productId);

  const originalPrice = product.discountPercentage
    ? product.price * (1 + product.discountPercentage / 100)
    : 0;

  const stockLevel =
    product.stock !== undefined
      ? product.stock > 20
        ? "high"
        : product.stock > 5
        ? "medium"
        : product.stock > 0
        ? "low"
        : "out"
      : null;

  return (
    <>
      <div className="grid grid-cols-1 md:grid-cols-2 gap-8 p-6">
        <ProductInfoGallery
          thumbnail={product.thumbnail}
          title={product.title}
          discountPercentage={product.discountPercentage}
        />

        <div className="flex flex-col">
          <ProductInfoHeader
            brand={product.brand}
            title={product.title}
            rating={product.rating}
            stockLevel={stockLevel}
            stock={product.stock}
          />

          <ProductInfoPricing
            price={product.price}
            originalPrice={originalPrice}
            discountPercentage={product.discountPercentage}
          />

          <ProductInfoDetails
            description={product.description}
            tags={product.tags}
          />

          <ProductInfoActions />

          <ProductInfoShipping />
        </div>
      </div>

      <div className="px-6 pb-6">
        <ProductInfoTabs product={product} />
      </div>
    </>
  );
};

export default ProductInfo;
