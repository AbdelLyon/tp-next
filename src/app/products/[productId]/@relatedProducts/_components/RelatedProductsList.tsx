// RelatedProductsList.tsx
import { ProductModel } from "@/types/product";
import { RelatedProductCard } from "./RelatedProductCard";

interface RelatedProductsListProps {
  products: ProductModel[];
}

export const RelatedProductsList = ({ products }: RelatedProductsListProps) => {
  return (
    <>
      {products.length > 0 ? (
        <div className="space-y-3" data-testid="related-products-list">
          {products.map((product) => (
            <RelatedProductCard key={product.id} product={product} />
          ))}
        </div>
      ) : (
        <div
          className="text-center py-10 rounded-md bg-gray-50/50 border border-dashed border-gray-200"
          data-testid="empty-products-message"
        >
          <p className="text-muted-foreground">No related products found</p>
        </div>
      )}
    </>
  );
};
