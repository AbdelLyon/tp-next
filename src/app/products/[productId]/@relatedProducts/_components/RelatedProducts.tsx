// RelatedProducts.tsx
import { productService } from "@/services/ProductService";
import { ChevronRight } from "lucide-react";
import { Separator } from "@/components/ui/separator";
import { ProductModel } from "@/types/product";
import { RelatedProductsList } from "./RelatedProductsList";
import { RelatedProductsHeader } from "./RelatedProductsHeader";
import { RelatedtCategoryList } from "./RelatedtCategoryList";

const RelatedProducts = async ({ productId }: { productId: string }) => {
  // Fonction isolée pour récupérer les produits associés
  // Cette fonction peut être facilement mockée pour les tests
  async function getRelatedProductsData(
    id: string,
  ): Promise<Array<ProductModel>> {
    const data = await productService.getProducts();
    return data.products
      .filter((product) => product.id !== Number(id))
      .slice(0, 4);
  }

  const relatedProducts = await getRelatedProductsData(productId);

  return (
    <>
      <RelatedProductsHeader
        title="You may also like"
        actionLink="/products"
        actionText="View all"
        icon={<ChevronRight className="h-3 w-3" />}
      />

      <RelatedProductsList products={relatedProducts} />

      <Separator className="my-6 bg-gradient-to-r from-transparent via-gray-200 to-transparent" />

      <RelatedtCategoryList
        title="Popular Categories"
        categories={["Electronics", "Smartphones", "Laptops", "Beauty"]}
      />
    </>
  );
};

export default RelatedProducts;
