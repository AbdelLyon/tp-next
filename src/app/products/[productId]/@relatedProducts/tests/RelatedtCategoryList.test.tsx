import { render, screen } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";
import { RelatedtCategoryList } from "../_components/RelatedtCategoryList";
vi.mock("@/components/ui/badge", () => ({
  Badge: ({
    children,
    className,
    "data-testid": dataTestId,
    ...rest
  }: {
    children: React.ReactNode;
    className?: string;
    "data-testid"?: string;
    variant?: string;
  }) => (
    <div className={className} data-testid={dataTestId} {...rest}>
      {children}
    </div>
  ),
}));
describe("<RelatedtCategoryList />", () => {
  // Test snapshot
  it("should match snapshot", () => {
    const { container } = render(
      <RelatedtCategoryList
        title="Popular Categories"
        categories={["Electronics", "Home Decor", "Fashion"]}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  // Test du rendu de base
  it("should render the component with title", () => {
    render(
      <RelatedtCategoryList
        title="Popular Categories"
        categories={["Electronics", "Home Decor", "Fashion"]}
      />,
    );

    // Vérifie que le conteneur principal est présent
    expect(screen.getByTestId("category-list")).toBeInTheDocument();

    // Vérifie que le titre est correctement affiché
    expect(screen.getByTestId("category-list-title")).toHaveTextContent(
      "Popular Categories",
    );
  });

  // Test du rendu des catégories
  it("should render all categories as badges", () => {
    const categories = ["Electronics", "Home Decor", "Fashion"];

    render(
      <RelatedtCategoryList
        title="Popular Categories"
        categories={categories}
      />,
    );

    // Vérifie que chaque catégorie est rendue
    expect(screen.getByTestId("category-electronics")).toBeInTheDocument();
    expect(screen.getByTestId("category-home-decor")).toBeInTheDocument();
    expect(screen.getByTestId("category-fashion")).toBeInTheDocument();

    // Vérifie que le texte de chaque badge correspond à la catégorie
    expect(screen.getByTestId("category-electronics")).toHaveTextContent(
      "Electronics",
    );
    expect(screen.getByTestId("category-home-decor")).toHaveTextContent(
      "Home Decor",
    );
    expect(screen.getByTestId("category-fashion")).toHaveTextContent("Fashion");
  });

  // Test avec une liste vide de catégories
  it("should not render any badges when categories array is empty", () => {
    render(<RelatedtCategoryList title="Popular Categories" categories={[]} />);

    // Vérifie que le conteneur principal est présent
    expect(screen.getByTestId("category-list")).toBeInTheDocument();

    // Vérifie qu'aucun badge n'est rendu
    const badgeContainer = screen
      .getByTestId("category-list")
      .querySelector(".flex.flex-wrap");
    expect(badgeContainer).toBeEmptyDOMElement();
  });

  // Test avec un grand nombre de catégories
  it("should handle a large number of categories", () => {
    const manyCategories = Array.from(
      { length: 10 },
      (_, i) => `Category-${i}`,
    );

    render(
      <RelatedtCategoryList
        title="Popular Categories"
        categories={manyCategories}
      />,
    );

    // Vérifie que toutes les catégories sont rendues
    manyCategories.forEach((category) => {
      const testId = `category-category-${category.toLowerCase().slice(9)}`;
      expect(screen.getByTestId(testId)).toBeInTheDocument();
    });
  });

  // Test des classes CSS appliquées
  it("should have correct CSS classes", () => {
    render(
      <RelatedtCategoryList
        title="Popular Categories"
        categories={["Electronics"]}
      />,
    );

    // Vérifie les classes du conteneur principal
    expect(screen.getByTestId("category-list")).toHaveClass(
      "flex flex-col space-y-3",
    );

    // Vérifie les classes du titre
    expect(screen.getByTestId("category-list-title")).toHaveClass(
      "text-sm font-medium",
    );

    // Vérifie les classes du badge
    expect(screen.getByTestId("category-electronics")).toHaveClass(
      "cursor-pointer hover:bg-primary/10 hover:text-primary transition-all duration-300 py-1.5",
    );
  });

  // Test pour vérifier que les caractères spéciaux dans les noms de catégories sont correctement gérés
  it("should handle special characters in category names", () => {
    const specialCategories = [
      "Kids & Baby",
      "Home & Garden",
      "Sports & Outdoors",
    ];

    render(
      <RelatedtCategoryList
        title="Special Categories"
        categories={specialCategories}
      />,
    );

    // Vérifie que les catégories avec des caractères spéciaux sont correctement rendues
    // Utiliser les data-testid tels qu'ils apparaissent réellement dans le DOM
    expect(screen.getByTestId("category-kids-&-baby")).toBeInTheDocument();
    expect(screen.getByTestId("category-home-&-garden")).toBeInTheDocument();
    expect(
      screen.getByTestId("category-sports-&-outdoors"),
    ).toBeInTheDocument();
  });
});
