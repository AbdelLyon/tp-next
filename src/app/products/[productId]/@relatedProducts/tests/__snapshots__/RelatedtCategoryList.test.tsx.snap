// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<RelatedtCategoryList /> > should match snapshot 1`] = `
<div>
  <div
    class="flex flex-col space-y-3"
    data-testid="category-list"
  >
    <h3
      class="text-sm font-medium"
      data-testid="category-list-title"
    >
      Popular Categories
    </h3>
    <div
      class="flex flex-wrap gap-2"
    >
      <div
        class="cursor-pointer hover:bg-primary/10 hover:text-primary transition-all duration-300 py-1.5"
        data-testid="category-electronics"
        variant="secondary"
      >
        Electronics
      </div>
      <div
        class="cursor-pointer hover:bg-primary/10 hover:text-primary transition-all duration-300 py-1.5"
        data-testid="category-home-decor"
        variant="secondary"
      >
        Home Decor
      </div>
      <div
        class="cursor-pointer hover:bg-primary/10 hover:text-primary transition-all duration-300 py-1.5"
        data-testid="category-fashion"
        variant="secondary"
      >
        Fashion
      </div>
    </div>
  </div>
</div>
`;
